### PR TITLE
[#6740] tests for iqstat (4-2-stable)

### PR DESCRIPTION
**Pull request by d-w-moore**
To go along with https://github.com/irods/irods_client_icommands/pull/349

### CODE REVIEW

**Comment by korydraughn** [comment]
Is this ready for review?

**Review by alanking — COMMENTED** [review]
Please mark as ready for review if true

File: scripts/irods/test/test_iqstat.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    def test_iqstat_with_all_option_and_no_id__6740(self):
+
+        rule_ids = [ self.spawn_delay_rule(self.user),
+                     self.spawn_delay_rule(self.admin) ]
+
+        # Assert brief format
+        _,out,err = self.admin.assert_icommand('iqstat -a','STDOUT_SINGLELINE', r'id\s\s*name', use_regex = True)
+        for rule_id in rule_ids:
+            self.assertRegexpMatches(out, re.compile('^' + rule_id + r'\s\s*',re.MULTILINE))
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:60]: We're searching for each of the rule IDs to appear in the multiline `STDERR` output string,  on different lines, like a table, which is appropriate for the kind of output iqstat produces when no id is given.

File: scripts/irods/test/test_iqstat.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    def tearDown(self):
+        for rule_id in self.rule_ids.split():
+            self.admin.assert_icommand('iqdel {}'.format(rule_id),'EMPTY')
+        super(Test_Iqstat, self).tearDown()
+
+    def spawn_delay_rule(self,sess):
+        dtnow = datetime.datetime.now()
+        epoch_with_micros = '({0:%s})({0:%f})'.format(dtnow)
+        sess.assert_icommand('''irule -r irods_rule_engine_plugin-irods_rule_language-instance '''
+                             '''"delay('<EF>1m</EF><PLUSET>1s</PLUSET>'){{'{}'}}" null null'''
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:38]: good point.

File: scripts/irods/test/test_iqstat.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+if sys.version_info < (2, 7):
+    import unittest2 as unittest
+else:
+    import unittest
+
+from . import command
+from . import session
+from .. import lib
+from .. import test
+from .. import paths
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:17]: Yes, I believe so.

File: scripts/irods/test/test_iqstat.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:38]: They are repeating rules (EF is specified) so I suppose there's no chance of them slipping out of the queue unless they fail, right?  But maybe you are correct, it's a more simple and direct presentation to make the PLUSET time greater....

[the remark this answers is not in the document]
Comment (alanking) [reply scripts/irods/test/test_iqstat.py:34]: Ah, I didn't see the `EF`. If that ensures that the rule repeats forever, then I'm good with how ti is. Please resolve if true :)

File: scripts/irods/test/test_iqstat.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alanking) [reply scripts/irods/test/test_iqstat.py:56]: I was more referring to the fact that this regex string is different from the one on 50, 58, 64, and 68. Wasn't sure if it was on purpose or if some parts were erroneously elided

[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:56]: Right - pretty much it's the difference between scanning for a bare id as first thing on each line, and scanning for the other specific patterns in the long format, ie records including something like id:\s\s*{ID} separated by '---'

File: scripts/irods/test/test_iqstat.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:34]: ok, cool... will do

File: scripts/irods/test/test_iqstat.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alanking) [reply scripts/irods/test/test_iqstat.py:56]: Okay, sounds good. Resolving.

File: scripts/irods/test/test_iqstat.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    def spawn_delay_rule(self,sess):
+        dtnow = datetime.datetime.now()
+        epoch_with_micros = '({0:%s})({0:%f})'.format(dtnow)
+        sess.assert_icommand('''irule -r irods_rule_engine_plugin-irods_rule_language-instance '''
+                             '''"delay('<EF>1m</EF><PLUSET>1s</PLUSET>'){{'{}'}}" null null'''
+                             ''.format(epoch_with_micros), 'EMPTY', desired_rc = 0)
+        rc, rule_id, err = sess.assert_icommand('''iquest "%s" "select RULE_EXEC_ID where RULE_EXEC_NAME like '%{}%'"'''
+                                                    ''.format(epoch_with_micros),'STDOUT',desired_rc = 0)
+        self.rule_ids += rule_id
```
Comment (korydraughn) [inline scripts/irods/test/test_iqstat.py:38]: You're concatenating strings here. Is a space separating them?

You call `spawn_delay_rule` twice in one test, so I'd expect this line to result in something like `rule_id_1rule_id_2`. The tear down code seems to expect the ids to be delimited in some way.

Am I misunderstanding?

**Review by korydraughn — COMMENTED** [review]
There's some inconsistency with how commas and whitespace are handled.

Consider adding a space after commas for readability. For example:
```python
foo(x,y,z) -> foo(x, y, z)
```

File: scripts/irods/test/test_iqstat.py (hunk 4 above)
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:38]: `self.rule_ids` actually ends up as `rule_id1\nrule_id2\n`; then in the teardown, `self.rule_ids.split( )`  produces ` [ 'rule_id1', 'rule_id2']`

Comment (korydraughn) [reply scripts/irods/test/test_iqstat.py:38]: So `+=` means `s = s + '\n' + t` where `s` and `t` are strings?

Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:38]: Each rule id returned from the command output includes its own terminating "\n".

Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py:38]: Maybe better if `rule_id `is rename` rule_id_with_carriage_return` ?

File: scripts/irods/test/test_iqstat.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    def setUp(self):
+        super(Test_Iqstat, self).setUp()
+        self.rule_ids = ''
+        self.admin = self.admin_sessions[0]
+        self.user = self.user_sessions[0]
+
+    def tearDown(self):
+        for rule_id in self.rule_ids.split():
+            self.admin.assert_icommand('iqdel {}'.format(rule_id),'EMPTY')
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py]: Done!

File: scripts/irods/test/test_iqstat.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        epoch_with_micros = '({0:%s})({0:%f})'.format(dtnow)
+        sess.assert_icommand('''irule -r irods_rule_engine_plugin-irods_rule_language-instance '''
+                             '''"delay('<EF>1m</EF><PLUSET>1s</PLUSET>'){{'{}'}}" null null'''
+                             ''.format(epoch_with_micros), 'EMPTY', desired_rc = 0)
+        rc, rule_id, err = sess.assert_icommand('''iquest "%s" "select RULE_EXEC_ID where RULE_EXEC_NAME like '%{}%'"'''
+                                                    ''.format(epoch_with_micros),'STDOUT',desired_rc = 0)
+        self.rule_ids += rule_id
+        return rule_id.strip()
+
+    def test_iqstat_with_all_option_and_id__6740(self):
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply scripts/irods/test/test_iqstat.py]: Yep.

**Comment by d-w-moore** [comment]
> There's some inconsistency with how commas and whitespace are handled.
> 
> Consider adding a space after commas for readability. For example:
> 
> ```python
> foo(x,y,z) -> foo(x, y, z)
> ```

Taken care of now, I think.

**Comment by korydraughn** [comment]
And the tests pass?

**Comment by d-w-moore** [comment]
Yep, tested along with  https://github.com/irods/irods_client_icommands/pull/349

**Review by korydraughn — COMMENTED** [review]
I think we're good to go.

Please squash without the pounds and we'll take one more look before approving.

**Comment by d-w-moore** [comment]
squashed!

**Review by korydraughn — APPROVED** [review]
Let's add the pound.

**Comment by d-w-moore** [comment]
> Let's add the pound.

Added!

**Pull request merged**